include/exadg/solvers_and_preconditioners/multigrid/coarse_grid_solvers.h

```diff
@@ -58,7 +58,7 @@ class CoarseGridSolverBase
   update() = 0;
 };
 
-template<typename Operator>
+template<int dim, typename Operator>
```

Comment:
Is the additional template parameter necessary? I would prefer if it could be deduced from `Operator` - this also relates to your question of how to deduce the mapping type. Does it work to call `pde_operator.get_matrix_free().get_mapping_info().mapping` (which returns a pointer to a mapping). Alternatively, I would prefer to deduce the template parameter `dim` from `Operator::dimension` (to be introduced as a `static constexpr` in field as I believe we don't currently do it) or by templating the respective constructors on `dim` to be able to accept a mapping (I also did not check this). My main point is that `Operator` already contains a template on `dim`, so it feels it should give us the respective information.

Reply:
I agree. I did not mention this so far because having done the refactoring suggested in other comments, we might want to re-discuss which objects should be passed to the constructor of `PreconditionerML` and which should actually be initialized outside and then handed over to `PreconditionerML`. So my question is: what are the implications of having the data structures that require `dim` actually outside the class `PreconditionerML`?

Reply:
If i can use the `pde_operator.get_matrix_free().get_mapping_info()mapping` as Martin suggests, then we do not need the `dim` since I only need to `get_ML_parameter_list(ml_data, dim)`.
Let's see, I will ask Peter for the name of the branch to compile this and then push after testing.